### ISSUE TITLE
wait_for always waits for the full duration, except when under strace/gdb

### ISSUE THREAD

**Issue by griwes** [boostorg/process#69]
System: Debian sid, x86_64

Since 09381034276ce76eedd6936ae330159dec9938d9, the following program:

```
#include <iostream>
#include <string>
#include <vector>

#include <boost/process.hpp>

int main() {
    std::vector<std::string> args{"/bin/echo", "hello from child"};

    boost::process::ipstream out;

    auto begin = std::chrono::high_resolution_clock::now();
    bool timeout_flag = false;

    boost::process::child child(args);
    std::error_code ec;
    if (!child.wait_for(std::chrono::seconds{10}, ec)) {
        timeout_flag = true;
        child.terminate(ec);
    }

    std::cout << "after wait_for" << std::endl;
}
```

ends up taking 10 seconds to complete, even though the spawned process exits pretty much immediately... except when running under `strace` or `gdb`. When I attach `gdb` to an already running process, I can see that it is blocking in `sigtimedwait`.

**Comment by pfaffe** [comment]
Duplicate of #60?

**Comment by griwes** [comment]
@klemens-morgenstern have you been able to reproduce this? Or should I spend some more time investigating what may be the cause of this?

**Comment by pfaffe** [comment]
@klemens-morgenstern  It reproduces for me on 03cba8f.

**Comment by klemens-morgenstern** [comment]
Sure, but I can't reproduce it and the CI pass. So I am not sure how to approach this problem, honestly.

**Comment by griwes** [comment]
@klemens-morgenstern I don't have a good way to explain this yet, but as far as I can tell - the test you've linked verifies that the wait returns correct values after specific timeouts, and does some arbitrary test if a 1s timeout takes less than 5s, but that doesn't prove this bug isn't happening; what I described would still make that test pass.

I'm still able to reproduce this with 1.70.0.beta1; here's a docker image of that, built with Clang and libc++ from LLVM 8, that I've reproduced it with: `griwes/llvm-boost:8.0.0-1.70.0.beta1` (ubuntu-based, so use `apt` to install further packages you may need). Please just use `c++` to compile, it's configured to build with Clang and libc++. (Boost.Test isn't built in that image, in case that matters.)

Kernel versions of docker hosts I've reproduced this with, in case that matters:
* Linux <snip> 4.20.14-towo.1-siduction-amd64 #1 SMP PREEMPT siduction 4.20-26 (2019-03-05) x86_64 GNU/Linux
* (I'll also try on another machine with a more vanilla kernel later today)

**Comment by klemens-morgenstern** [comment]
Can you design a test that fails with the behaviour described by you?
Please note that this might be a Kernle issue, i.e. that the kernel implementes sigtimedwait differently.

**Comment by griwes** [comment]
Can't test right now, but I'd assume that something like this would fail (modeled after the first test in the test suite you've linked):

```
BOOST_AUTO_TEST_CASE(wait_for_exit_before_timeout)
{
    using boost::unit_test::framework::master_test_suite;

    std::error_code ec;

    auto launch_time = std::chrono::system_clock::now();
    bp::child c(
        master_test_suite().argv[1],
        bp::args+={"test", "--wait", "1"},
        ec
    );
    BOOST_REQUIRE(!ec);

    BOOST_CHECK( c.wait_for(std::chrono::seconds(20)));

    auto timeout_t = std::chrono::system_clock::now();

    // check that we didn't wait the entire timeout period
    BOOST_CHECK_LT(std::chrono::duration_cast<std::chrono::seconds>(timeout_t - launch_time).count(), 20);
}
```

The problem with this is that on a system that is sufficiently under load this could still _in theory_ fail even when the behavior is correct; I think that something like 20 seconds should be enough to be sure about it, but feel free to adjust it.

**Comment by klemens-morgenstern** [comment]
Alright, added. Waiting for results now, build's running [here](https://travis-ci.org/klemens-morgenstern/boost-process/builds/509265760).

**Comment by griwes** [comment]
I can see that that test has failed when you've run it initially. #197 does _not_ fix this, I can still see this same exact issue with current `develop`.

**Comment by senthil-ram** [comment]
Is this issue resolved? After upgrade to boost 1.72 - we are seeing this issue. Did not see it in boost 1.67.

**Comment by klemens-morgenstern** [comment]
I only added a test since 1.72, so the 1.73 wouldn't fix this. I will hopefully revisit the issue soon,

Are you running on OSX?

**Comment by klemens-morgenstern** [comment]
Would using `asio::deadline_timer` work for you? I am hesitant to do that implicitly, but that might be the only way to build a reliable solution.

**Comment by senthil-ram** [comment]
> I only added a test since 1.72, so the 1.73 wouldn't fix this. I will hopefully revisit the issue soon,
> 
> Are you running on OSX?

On linux centos6/7.

**Comment by emenchen** [comment]
Are you all sure that the process wasn't executing so quickly that it was done before you entered the wait? I'm not seeing this in 1.72. I put in a 30 second wait, and my test is finishing in ~10 seconds (doing other stuff). I'm checking running() before calling wait_for(std::chrono::seconds(30).

**Comment by jtousley** [comment]
@klemens-morgenstern Still seeing this issue as of Boost 1.81 - do we have any plans for a solution, or at least a workaround, yet? This was not a problem in Boost 1.68.

@emenchen I am checking running() before calling wait_for as in the example, and it occurs both when executing fast commands (echo) and slow commands (sleep).

**Comment by ivantishchenko** [comment]
@klemens-morgenstern @griwes 

I also get the same issue. In my use case, I start one boost process from a unit test using GTest:

```
    boost::process::ipstream pipeStream;
    boost::process::child boundingBoxChildProcess(
        boundingBoxCommand, boost::process::std_in.close(), boost::process::std_out.close(),
        boost::process::std_err > pipeStream);

     ...

     boundingBoxChildProcess.wait_for(timeout);

     ...
```

wait_for will block for the exact duration specified in the variable **timeout**.  Interestingly enough, If I substitute wait_for with wait the snippet will always take around 230 ms. If I put wait_for back with a timeout equal to 5 ms, the function will take ~ 5 seconds 230 ms.  As such, it's always timeout_durtaion + duration as with regular .wait().

Furthermore, I have another use case where I need to launch two boost processes one by one, with one process using the results of the previous process as inputs. If I use wait_for in both of these processes, sometimes the second process will be blocked indefinitely.  If I swap the call to wait_for(timeout) with wait() the code will be executed without waiting and no indefinite blocking will occur. Here are the outputs. In this run, I start two bounding box processes, followed by two rasterization processes, both type of processes use wait_for(timeout) to finish gracefully.
```
Starting to wait on bounding box
Starting to wait on bounding box
Bounding box waiting finished 
Bounding box waiting finished 
The sizes are 1 1
Starting to wait for rasterization

...

The execution hangs at this point
```
I assume that it's a special case of the previous use case.

Build information:
Boost version 1.74.0, gcc-11, compiled with C++20 on Debian GNU/Linux 12 (bookworm) x86_64, 6.1.0-15-amd64.

P.S.

I ended up using wait instead of wait_for and spawning a subprocess on a separate detached thread in my application. It solves the issue. However, it would be still nice to use wait_for instead to avoid potential zombie processes.

**Comment by ivantishchenko** [comment]
@emenchen 

> Are you all sure that the process wasn't executing so quickly that it was done before you entered the wait? I'm not seeing this in 1.72. I put in a 30-second wait, and my test is finishing in ~10 seconds (doing other stuff). I'm checking running() before calling wait_for(std::chrono::seconds(30).

I suspect that checking for `running()` isn't necessary before calling wait_for. wait_for employes wait_until which has a check for `if (!_exited())` inside.

**Comment by tstx** [comment]
```
CommandResult BoostProcessExecutor::execute(
    const std::string& program, const std::vector<std::string>& args,
    const std::string& input_data, const std::chrono::seconds& timeout )
{
    CommandResult result;
    bp::opstream in_stream;
    bp::ipstream out_stream;
    bp::ipstream error_stream;

    // clang-format off
    bp::child c( program, bp::args( args ),
                 bp::std_in  < in_stream,
                 bp::std_out > out_stream,
                 bp::std_err > error_stream );
    // clang-format on

    in_stream << input_data;
    in_stream.flush();
    in_stream.pipe().close();

    if ( !c.wait_for( timeout ) ) {
        c.terminate();
        result.timeout = true;
        return result;
    }

    std::string line;
    while ( std::getline( out_stream, line ) ) {
        result.stdout_data += line + "\n";
    }
    while ( std::getline( error_stream, line ) ) {
        result.stderr_data += line + "\n";
    }

    result.exit_code = c.exit_code();
    return result;
}
```

running into the same issue using
boost-devel-1.69.0-150400.1.1.noarch
  
```
TEST_F(CommandExecutorTest, EchoCommandReturnsCorrectOutput) {
    auto result = cmd_executor->execute("/bin/echo", 
        {"Hello World"}, "", std::chrono::seconds(1));
    
    EXPECT_EQ(result.exit_code, 0);
    EXPECT_EQ(result.stdout_data, "Hello World\n");
    EXPECT_FALSE(result.timeout);
}
```

[ RUN      ] CommandExecutorTest.EchoCommandReturnsCorrectOutput
[       OK ] CommandExecutorTest.EchoCommandReturnsCorrectOutput (1002 ms)

any there any updates regarding the bug?

**Comment by klemens-morgenstern** [comment]
The wait_for/wait_until functions are broken & deprecated. You should switch to boost process v2 and use asio's cancel_after.
